features/mbedtls/targets/TARGET_NXP/aes_alt.c

```diff
@@ -0,0 +1,609 @@
+    if (keybits == 256u)
+    {
+        return (MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE);
+    }
+#endif
+
+    ctx->rk = RK = ctx->buf;
+
+#if defined(MBEDTLS_FREESCALE_LTC_AES) || defined(MBEDTLS_FREESCALE_LPC_AES) || defined(MBEDTLS_FREESCALE_CAU3_AES) || \
+    defined(MBEDTLS_FREESCALE_CAAM_AES) || defined(MBEDTLS_FREESCALE_DCP_AES)
+    const unsigned char *key_tmp = key;
+    memcpy(RK, key_tmp, keybits / 8);
+
+#if defined(MBEDTLS_FREESCALE_CAU3_AES) || defined(MBEDTLS_FREESCALE_DCP_AES)
+    crypto_detach_ctx_from_key_slot(ctx);
+#endif /* MBEDTLS_FREESCALE_CAU3_AES || MBEDTLS_FREESCALE_DCP_AES */
+
+    switch (keybits)
+    {
+        case 128:
+            ctx->nr = 16;
+            break;
+        case 192:
+            ctx->nr = 24;
+            break;
+        case 256:
+            ctx->nr = 32;
+            break;
+        default:
+            return (MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
+    }
+#elif defined(MBEDTLS_FREESCALE_MMCAU_AES)
+    ctx->rk = RK = ctx->buf;
+
+    switch (keybits)
+    {
+        case 128:
+            ctx->nr = 10;
+            break;
+        case 192:
+            ctx->nr = 12;
+            break;
+        case 256:
+            ctx->nr = 14;
+            break;
+        default:
+            return (MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
+    }
+
+    MMCAU_AES_SetKey(key, keybits / 8, (uint8_t *)RK);
+#endif
+
+    return 0;
+}
+
+/*
+ * AES-ECB block encryption/decryption
+ */
+int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
+                    int mode,
+                    const unsigned char input[16],
+                    unsigned char output[16] )
+{
+#if defined(MBEDTLS_AESNI_C) && defined(MBEDTLS_HAVE_X86_64)
+    if (mbedtls_aesni_has_support(MBEDTLS_AESNI_AES))
+        return (mbedtls_aesni_crypt_ecb (ctx, mode, input, output));
+#endif
+
+#if defined(MBEDTLS_PADLOCK_C) && defined(MBEDTLS_HAVE_X86)
+    if (aes_padlock_ace) {
+        if (mbedtls_padlock_xcryptecb (ctx, mode, input, output) == 0)
+            return (0);
+
+        // If padlock data misaligned, we just fall back to
+        // unaccelerated mode
+        //
+    }
+#endif
+
+    if (mode == MBEDTLS_AES_ENCRYPT) {
+        return (mbedtls_internal_aes_encrypt(ctx, input, output));
+    } else {
+        return (mbedtls_internal_aes_decrypt(ctx, input, output));
+    }
+
+    return( 0 );
+}
+
+/*
+ * AES-ECB block encryption
+ */
+int mbedtls_internal_aes_encrypt(mbedtls_aes_context *ctx, const unsigned char input[16], unsigned char output[16])
+{
+    uint8_t *key;
+
+    key = (uint8_t *)ctx->rk;
+#if defined(MBEDTLS_FREESCALE_LTC_AES)
+    LTC_AES_EncryptEcb(LTC_INSTANCE, input, output, 16, key, ctx->nr);
+#elif defined(MBEDTLS_FREESCALE_MMCAU_AES)
+    MMCAU_AES_EncryptEcb(input, key, ctx->nr, output);
+#elif defined(MBEDTLS_FREESCALE_CAU3_AES)
+    if (!crypto_key_is_loaded(ctx))
+    {
+        CAU3_AES_SetKey(CAU3, &s_cau3Handle, key, ctx->nr);
+        crypto_attach_ctx_to_key_slot(ctx, s_cau3Handle.keySlot);
```

Comment:
Is there any reason it is preferred to lazy load the key into the accelerator like this rather than load it explicitly in mbedtls_aes_setkey_enc()? If so, please add a C comment explaining your reasoning.

Reply:
I have deleted CAU3 code as this is not needed for the platforms supported by this PR